### PR TITLE
cql3: Optimize LIKE regex construction

### PR DESCRIPTION
**Pull request by dekimir**
Currently we create a regex from the LIKE pattern for every row
considered during filtering, even though the pattern is always the
same.  This is wasteful, especially since we require costly
optimization in the regex compiler.  Fix it by reusing the regex
whenever the pattern is unchanged since the last call.

Tests: unit (dev)

Signed-off-by: Dejan Mircevski <dejan@scylladb.com>

### CODE REVIEW

File: cql3/restrictions/single_column_restriction.hh [hunk 1]
```diff
@@ -384,6 +387,11 @@ public:
 class single_column_restriction::LIKE final : public single_column_restriction {
 private:
     ::shared_ptr<term> _value;
+    /// Matches cell value against LIKE pattern.  Pointer because it cannot be initialized in the
+    /// constructor when the pattern is a bind marker.  Mutable because it is initialized on demand
+    /// in is_satisfied_by().
+    mutable std::unique_ptr<like_matcher> _matcher;
```
Comment (denesb) [inline cql3/restrictions/single_column_restriction.hh]: `std::optional` is another option here, we usually use `std::unique_ptr` when stable address is needed.

File: cql3/restrictions/statement_restrictions.cc [hunk 2]
```diff
@@ -916,6 +915,18 @@ bool token_restriction::slice::is_satisfied_by(const schema& schema,
...
 }
 
+bool single_column_restriction::LIKE::init_matcher(const query_options& options) const {
+    auto pattern = to_bytes_opt(_value->bind_and_get(options));
+    if (!pattern) {
+        return false;
+    }
+    if (!_matcher || pattern != _last_pattern) {
+        _matcher.reset(new like_matcher(*pattern));
+        _last_pattern = std::move(pattern);
```
Comment (avikivity) [inline cql3/restrictions/statement_restrictions.cc]: std::make_unique<> is better practice than a naked new. std::optional<> would be even better.

File: tests/cql_query_test.cc [hunk 3]
```diff
@@ -3722,6 +3722,21 @@ void require_rows(cql_test_env& e,
...
+void require_rows(cql_test_env& e,
+                  cql3::prepared_cache_key_type id,
+                  const std::vector<cql3::raw_value>& values,
+                  const std::vector<std::vector<bytes_opt>>& expected,
+                  const source_location& loc = source_location::current()) {
+    try {
+        assert_that(e.execute_prepared(id, values).get0()).is_rows().with_rows_ignore_order(expected);
+    }
+    catch (const std::exception& e) {
+        BOOST_FAIL(format("execute_prepared failed: {}\n{}:{}: originally from here",
```
Comment (avikivity) [inline tests/cql_query_test.cc]: Our coding style is  `} catch (` in one line.

**Comment by dekimir** [comment]
Feedback addressed, PTAL.

**Pull request merged**